keps/sig-network/3458-remove-transient-node-predicates-from-service-controller/README.md

```diff
@@ -0,0 +1,337 @@
+  do however support this, and it is not expected that other providers don't.
+- By removing the b) predicate above we are delaying the removal of the node
+  from the load balancers' node set until the node is completely deleted in the
+  API server. Workloads relying on this behavior will observe a change in
+  behavior, but this also not expected.
+
+#### Mitigations
+
+- Cloud providers/workloads which do not support the behavior mentioned in
+  (Risk)[#risk], have the possibility to set the feature flag to false, thus
+  default back to the current mechanism.
+
+## Design Details
+
+- Implement the change in the service controller and ensure it does not add /
+  remove nodes from the load balancers' node set for cases a) and b) mentioned
+  in (Summary)[#Summary]
+- Add the feature gate: `LoadBalancerConnectionDraining`, set it to "on" by
+  default and promote it directly to GA
+
+### Test Plan
+
+#### Prerequisite testing updates
+
+#### Unit tests
+
+The service controller in the KCCM currently has a set of tests validating
+expected syncs caused by Node predicates, these will need to be updated.
+
+#### Integration tests
+
+Kubernetes is mostly tested via unit tests and e2e, not integration, and this is
+not expected to change.
+
+#### e2e tests
+
+Kubernetes in general needs to extended its load balancing test suite with
+disruption tests, this might be the right effort we need to get that ball
+rolling. Testing would include:
+
+- validation that an application running on a deleting VM benefits from graceful
+  termination of its TCP connection.
+- validation that Node readiness state changes to not result in load balancer
+  re-syncs.
+
+### Graduation Criteria
+
+#### GA
+
+This is addressing a sub-optimal solution currently existing in Kubernetes, so
+the feature gate will be GA and "on" by default from the start.
+
+The feature flag should be kept available until we get sufficient evidence of
+people not being affected by this - by not being able to delete VMs on
+downscaling by the cluster auto-scaler.
+
+Tentative timeline is to lock the feature flag to true in 1.27.
+
+### Upgrade / Downgrade Strategy
+
+N/A
+
+### Version Skew Strategy
+
+N/A
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: `LoadBalancerConnectionDraining`
+  - Components depending on the feature gate: Kubernetes cloud controller manager
+
+###### Does enabling the feature change any default behavior?
+
+Yes, Kubernetes Nodes will remain in the load balancers' node set until fully
+deleted in the API server, as opposed to the current behavior: which adds /
+removes the nodes from the set when the Node experience transient state changes.
+Cloud providers which do not support deleting VMs which are still referenced by
```

Comment:
If CA taints a Node, then deletes the VM, and NodeController is slow (say it takes 60 seconds to wake up), any subsequent LB set changes will include this Node, but the VM for it is already gone, so will fail.  Should we also encourage implementations to check that the VMs exist before adding them?

Reply:
> Should we also encourage implementations to check that the VMs exist before adding them?

Ah yeah: forgot to mention this. Will add now.

How do we "encourage" them? That will most likely force them to perform a `cloudprovider.InstanceExists(node)` call for all nodes that the service controller provides at each sync, which adds to the amount of cloud API calls, and which they won't be very keen on. The positive part is that we will only sync if another node is annotated with "exclude" / added / removed: so we will sync much less. If we don't manage to encourage them, then the service controller will error for a while until we reach steady-state (acceptable, I think)

Reply:
I added this to `Risk`

Reply:
will comment there